pkg/target/target.go

```diff
@@ -426,6 +430,65 @@ func convertToLabelSelector(object map[string]interface{}) (*metav1.LabelSelecto
 	return obj, nil
 }
 
+// TODO: can we use generic for Unmarshal after go 1.18?
+func convertToMatch(object map[string]interface{}) (*match.Match, error) {
+	j, err := json.Marshal(object)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not convert unknown object to JSON")
+	}
+	obj := &match.Match{}
+	if err := json.Unmarshal(j, obj); err != nil {
+		return nil, errors.Wrap(err, "Could not convert JSON to Match")
+	}
+	return obj, nil
+}
+
+// ToMatcher converts .spec.match in mutators to Matcher.
 func (h *K8sValidationTarget) ToMatcher(u *unstructured.Unstructured) (constraints.Matcher, error) {
-	panic("not implemented")
+	obj, found, err := unstructured.NestedMap(u.Object, "spec", "match")
+	if err != nil {
+		return nil, err
+	}
+	if found && obj != nil {
+		match, err := convertToMatch(obj)
+		if err != nil {
+			return nil, err
+		}
+		return &Matcher{match}, nil
+	}
+	return nil, fmt.Errorf("%s %s has no match found", u.GetKind(), u.GetName())
 }
+
+type Matcher struct {
+	match *match.Match
+}
+
+func (m *Matcher) Match(review interface{}) (bool, error) {
```

Comment:
@willbeason It occurs to me that some matchers are applied against both `object` and `oldObject` in order to prevent users from bypassing constraints by removing a relevant label:

https://github.com/open-policy-agent/gatekeeper/blob/9a02fe8b3fbd336c1178cb393d987741a914ae7a/pkg/target/regolib/src.rego#L276-L285

It doesn't look like the mutation matcher does this. Maybe we can work around it by calling match() for both `object` and `oldObject`?

Reply:
You're right - I didn't know Gatekeeper did that. Fortunately, this is automatically filled in for the Admission request.

Reply:
Updated match for both object and oldObject.